Added Rosetta's construction/combine endpoint.

cngo-github opened the pull request:
__PR Category__: Feature

Added Rosetta's [`construction/combine` endpoint](https://www.rosetta-api.org/docs/ConstructionApi.html#constructioncombine).

cngo-github commented:
I want to get the return type here:

https://github.com/cngo-github/tessellation/blob/20230310-construction_combine_redux/modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala#L121-L150

to `EitherT[F, ConstructionError, Hex]`, but I'm not sure how.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+        }
+        .attemptT
+
+    def convertSignature(
+      signature: NonEmptyList[RosettaSignature]
+    ) = {
```
use plural name as it takes a list of signatures, not a single signature

cngo-github replied:
Changed to `signatures`.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
         parseUnsignedTransaction(hex)
       }
+
+    def combineTransaction(hex: Hex, signatures: NonEmptyList[RosettaSignature]): EitherT[F, Throwable, Either[ConstructionError, Hex]] =
```
these two methods written in a functional way could be dramatically simplified. also, method for converting signatures is overengineered and the traversing is obfuscated. it's recommended to use traversing provided by cats:
```scala
    def combineTransaction(hex: Hex, signatures: NonEmptyList[RosettaSignature]): EitherT[F, ConstructionError, Hex] =
      EitherT
        .fromEither(KryoSerializer[F].deserialize[Transaction](hex.toBytes))
        .leftMap(_ => MalformedTransaction)
        .flatMap { transaction =>
          convertSignatures(signatures).flatMap { signatures =>
            EitherT
              .fromEither(
                KryoSerializer[F]
                  .serialize(Signed[Transaction](transaction, NonEmptySet(signatures.head, SortedSet(signatures.tail: _*))))
              )
              .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
              .map(Hex.fromBytes(_))
          }
        }

    def convertSignatures(signatures: NonEmptyList[RosettaSignature]): EitherT[F, ConstructionError, NonEmptyList[SignatureProof]] =
      EitherT {
        signatures.traverse { signature =>
          signature.publicKey.hexBytes.toPublicKeyByEC.map { pk =>
            signature.signatureType match {
              case SignatureType.Ecdsa => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError]
              case _                   => UnsupportedOperation.asInstanceOf[ConstructionError].asLeft[SignatureProof]
            }
          }
        }.map(_.sequence)
      }

```

cngo-github replied:
So, use `traverse` instead of `foldLeft`?

cngo-github replied:
I've implemented the functions as shown, but I will need to study them more to understand them.

cngo-github replied:
The return type for `combineTransaction` is still `EitherT[F, Throwable, Either[ConstructionError, Hex]]`. I would like it to be `EitherT[F, ConstructionError, Hex]`. I think we did a `liftTo` in a previous, similar situation, but I don't completely understand lifting and `liftTo`. Am I looking in the right direction here?

marcinwadon replied:
What do you mean that the return type is still `EitherT[F, Throwable, Either[ConstructionError, Hex]]`?
It's of type `EitherT[F, ConstructionError, Hex]`, you can see that clearly here:
![image](https://user-images.githubusercontent.com/4959931/224950825-54823fb6-3b16-4d04-9884-04369f042ecb.png)

cngo-github replied:
My mistake on the return type. The error was on my side.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+        }
+      }
+      val zero: Either[ConstructionError, List[SignatureProof]] = Right[ConstructionError, List[SignatureProof]](List())
+      value
+        .foldLeft(Async[F].delay(zero)) {
+          case (agga, nextFEither) =>
+            val inner2 = agga.flatMap { agg =>
+              val inner = nextFEither.map { eitherErrSigProof =>
+                val res = eitherErrSigProof.map { sp =>
+                  agg.map(ls => ls ++ List(sp))
+                }
+                var rett: Either[ConstructionError, List[SignatureProof]] = null
+                if (res.isRight) {
+                  val value2 = res.toOption.get
+                  val zz = value2.getOrElse(List())
+                  if (zz.nonEmpty) {
+                    rett = Right(zz)
+                  } else {
+                    rett = Left(value2.swap.toOption.get)
+                  }
+                } else {
+                  rett = Left(res.swap.toOption.get)
+                }
+                rett
+              }
+              inner
+            }
+            inner2
+        }
+        .map(_.map(q => NonEmptyList(q.head, q.tail)))
```
it's very error prone and actually obfuscated what it does. if you start doing things such way, think twice or ask, because most likely it can be written in way simpler

cngo-github replied:
Can you elucidate on what is bad here?

marcinwadon replied:
First of all, you use mutations which should be avoided as much as possible.
Secondly, there are a lot of auxiliary variables that are not necessary. Converting from Option to Either is overengineered, you don't need to use `if`. You used `.get` operation on option which is unsafe and can throw in the runtime. Overall, it looks very imperative and it becomes unmaintainable really quickly. 

There are several reasons why declarative programming is often considered better and easier to understand than imperative programming:

- Readability: Declarative code is often more readable because it focuses on the intent of the code rather than the implementation details. This makes it easier for other developers to understand and maintain the code.

- Simplicity: Declarative code is often simpler and more concise than imperative code because it abstracts away many of the low-level details. This can lead to faster development times and fewer bugs.

- Reusability: Declarative code is often more reusable than imperative code because it can be used in a wider variety of contexts. This can lead to more modular and maintainable code.

- Parallelism: Declarative programming is often more amenable to parallelization because it specifies the desired outcome rather than the exact steps needed to achieve it. This can lead to faster execution times on multi-core or distributed systems.

Overall, declarative programming is often seen as a more natural and intuitive way of expressing complex computations because it allows the programmer to focus on the what rather than the how.

Instead of thinking **how** to do something, start thinking **what** to do.

**As an example:**

I have a non empty list of signatures. I want to convert each signature from rosetta schema to SignatureProof. 

```scala
    def convertSignatures(signatures: NonEmptyList[RosettaSignature]): EitherT[F, ConstructionError, NonEmptyList[SignatureProof]] = ???
```

I know that for each signature, I need to derive a public key:

```scala
signatures.traverse { signature =>
         signature.publicKey.hexBytes.toPublicKeyByEC.map { pk =>
         
         }
}
```

then, based on the rosetta specification, I want only ECDSA signatures to be converted:

```scala
            signature.signatureType match {
              case SignatureType.Ecdsa => ???
              case _                   => ???
            }
```

For ECDSA signature type, I want to create a SignatureProof, for other cases I want to return an unsupported operation.

```scala
         case SignatureType.Ecdsa => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError]
         case _                   => UnsupportedOperation.asInstanceOf[ConstructionError].asLeft[SignatureProof]
```

I return it wrapped into an Either (to have a common type), so I end up with type:

```scala
F[NonEmptyList[Either[ConstructionError, Hex]]]
```

I need sequence operation to convert from `F[G[A]]` into `G[F[A]]`
```scala
.map(_.sequence)
```

so it becomes:
```scala
F[Either[ConstructionError, NonEmptyList[Hex]]]
```

Then, I want to convert `F[Either` into `EitherT[F`, so I need to wrap it into `EitherT { ... }`.
Done.

marcinwadon replied:
No, traverse **is** the map then sequence. (Take a look info docs and traverse/sequence implementations)

marcinwadon replied:
or actually @cngo-github is the opposite, `sequence` **is** the `traverse` but with identity map function:
![image](https://user-images.githubusercontent.com/4959931/225120809-cd824a7f-1d76-49e0-8d47-8d2369c3abe9.png)
![image](https://user-images.githubusercontent.com/4959931/225120836-842bf45a-67b6-452c-a4b5-ad04183b2d53.png)

marcinwadon replied:
also this @cngo-github 
![image](https://user-images.githubusercontent.com/4959931/225122404-6a20014e-5ed9-4a94-8e0d-82425eb33744.png)

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/RosettaSignature.scala:
```diff
@@ -0,0 +1,34 @@
+object SignatureType extends StringEnum[SignatureType] with StringCirceEnum[SignatureType] {
+  val values = findValues
+
+  case object Ecdsa extends SignatureType(value = "ecdsa")
```
Changed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/api/construction/ConstructionCombine.scala:
```diff
@@ -0,0 +1,23 @@
+  @derive(customizableDecoder)
+  case class Request(
+    networkIdentifier: NetworkIdentifier,
+    unsigned_transaction: Hex,
```
Fixed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+        .map(_.map(q => NonEmptyList(q.head, q.tail)))
+    }
+
+    // This works from:
+    //  https://stackoverflow.com/questions/26159149/how-can-i-get-a-publickey-object-from-ec-public-key-bytes
+    def getPublicKeyFromBytes(pubKey: Array[Byte]) = {
+      val spec = ECNamedCurveTable.getParameterSpec("secp256k1")
+      val kf = KeyFactory.getInstance("ECDSA", new BouncyCastleProvider)
+      val params = new ECNamedCurveSpec("secp256k1", spec.getCurve, spec.getG, spec.getN)
+      val point = ECPointUtil.decodePoint(params.getCurve, pubKey)
+      val pubKeySpec = new ECPublicKeySpec(point, params)
+      kf.generatePublic(pubKeySpec).asInstanceOf[ECPublicKey]
+    }
```
I talked about our discussion with @ryle-goehausen . He tells me that the implemented method produces a key of a different format than what is needed. So, this method is needed. I seem to remember discussing with you a better way to do this, but I have forgotten the specifics of our discussion. I'll review the recordings of our various discussions and previous PRs to see if that was captured.

marcinwadon replied:
I think you're wrong @cngo-github. It returns exactly the same public key spec

marcinwadon replied:
> He tells me that the implemented method produces a key of a different format than what is needed.
@cngo-github can you verify that first?

cngo-github replied:
I talked with @ryle-goehausen , the `toPublicKeyByEC` is the implementation I will use.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala:
```diff
@@ -57,6 +59,22 @@ final class ConstructionRoutes[F[_]: Async](
+      req
+        .decodeRosettaWithNetworkValidation[ConstructionCombine.Request](appEnvironment, _.networkIdentifier) { combineReq =>
+          NonEmptyList.fromList(combineReq.signatures).flatMap { signatures =>
+            if(signatures.length == 1) {
```
```
case signatures: RosettaSignature :: Nil => constructionService
                .combineTransaction(combineReq.unsigned_transaction, signatures).some
```

is giving me the error

```
'=>' expected but identifier found.
```

for the `::`

cngo-github replied:
Now, the error is

```
constructor cannot be instantiated to expected type;
 found   : scala.collection.immutable.::[A]
 required: cats.data.NonEmptyList[org.tessellation.rosetta.domain.RosettaSignature]
 ```

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+      KryoSerializer[F]
+        .deserialize[Transaction](hex.toBytes)
+        .traverse { transaction =>
+          convertSignature(signatures).map { test2 =>
```
Leftover from when I was debugging. changed to `_.left`

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+        .traverse { transaction =>
+          convertSignature(signatures).map { test2 =>
+            test2.left
+              .map(err => Left(err): Either[ConstructionError, Hex])
```
Changed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +118,92 @@ object ConstructionService {
+            test2.left
+              .map(err => Left(err): Either[ConstructionError, Hex])
+              .map { signatureProofs =>
+                val serializedTransaction: Either[ConstructionError, Hex] = KryoSerializer[F]
```
Changed.

cngo-github commented:
@marcinwadon , moving onto the route:

https://github.com/cngo-github/tessellation/blob/20230310-construction_combine_redux/modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala#L61-L79

there are two problems:

1. I am getting the error

```
constructor cannot be instantiated to expected type;
 found   : scala.collection.immutable.::[A]
 required: cats.data.NonEmptyList[org.tessellation.rosetta.domain.RosettaSignature]
```

on this [line](https://github.com/cngo-github/tessellation/blob/20230310-construction_combine_redux/modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala#L68).

2. I don't know how to deal with converting an `Option[EitherT[...]]` to `F[Response[F]]]`. So I am using a `get` [here](https://github.com/cngo-github/tessellation/blob/20230310-construction_combine_redux/modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala#L77), but that throws an exception and is not good in production code.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +111,35 @@ object ConstructionService {
+          signature.publicKey.hexBytes.toPublicKeyByEC.map { pk =>
+            signature.signatureType match {
+              case SignatureType.ECDSA => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError]
+              case _                   => UnsupportedOperation.asInstanceOf[ConstructionError].asLeft[SignatureProof]
```
FYI - I'm pretty sure this case would never happen, as the request would be rejected on the very first operation (trying to decode the request), as the ECDSA signature is the only one that is supported, so there is no possibility you can get any other type here (hence, match should have also a single element probably)

cngo-github replied:
The logic has been reworked and this function is no more.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/error/RosettaError.scala:
```diff
@@ -42,6 +42,13 @@ object RosettaError extends IntEnum[RosettaError] with RosettaErrorEncoder {
         retriable = false,
         description = Some("Unable to decode the transaction.")
       )
+  case object UnsupportedOperation
+      extends RosettaError(
+        value = 7,
+        message = "Unsupported operation",
+        retriable = false,
+        description = Some("The operation is not supported.")
+      )
```
according  to previous comment, this would never happen so most likely it's unused/unnecessary

cngo-github replied:
This is no used in the route when there are more than 1 signature.

marcinwadon commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/http/routes/ConstructionRoutes.scala:
```diff
@@ -57,6 +57,25 @@ final class ConstructionRoutes[F[_]: Async](
+            }
+            .map(_.bimap(_.toRosettaError, ConstructionCombine.Response(_)).asRosettaResponse.handleUnknownError)
+
+          response.get
```
if you want to use `get`, you should use `getOrElse` providing the value for empty (none) case.
anyway, you can type `combineReq.signatures` as NonEmptyList, so then you can just pass it to the construction service and map its output, without additional logic here in the routes

cngo-github replied:
I reworked the logic to not need this.

cngo-github commented:
This does not currently compile for the reasons in my last comment to @marcinwadon . I have added unit tests:

However, running the tests gives the following errors:

```
[31m  ClassCastException: class java.lang.Float cannot be cast to class org.tessellation.schema.transaction$Transaction (java.lang.Float is in module java.base of loader 'bootstrap'; org.tessellation.schema.transaction$Transaction is in unnamed module of loader 'app')[0m

[31m  EitherT.scala:446                     cats.data.EitherT#$anonfun$flatMap$1
[31m  ConstructionService.scala:49          leftMap @ org.tessellation.rosetta.domain.construction.ConstructionService$$anon$1#derive
[31m  ConstructionService.scala:119         flatMap @ org.tessellation.rosetta.domain.construction.ConstructionService$$anon$1#combineTransaction
[31m  ConstructionServiceSuite.scala:329    map @ org.tessellation.rosetta.domain.construction.ConstructionServiceSuite$#$anonfun$new$57
[31m  Test.scala:21                         defer @ weaver.Test$#$anonfun$apply$2
[31m  Stream.scala:2201                     get @ fs2.Stream#$anonfun$parEvalMapActionImpl$1
[31m  Test.scala:22                         map @ weaver.Test$#$anonfun$apply$2
[31m  Stream.scala:558                      voidError$extension @ fs2.Stream#$anonfun$concurrentlyAux$2
[31m  Test.scala:20                         flatMap @ weaver.Test$#$anonfun$apply$2
[31m  UnsafeRun.scala:13                    map @ weaver.EffectCompat#realTimeMillis
[31m  Test.scala:19                         flatMap @ weaver.Test$#$anonfun$apply$1
[31m  GlobalResourceF.scala:124             of$extension @ weaver.GlobalResourceF$#createMap
[31m  Test.scala:18                         flatMap @ weaver.Test$#apply
[31m  RunnerCompat.scala:180                release @ weaver.framework.RunnerCompat#$anonfun$run$12
[31m  Stream.scala:2185                     flatMap @ fs2.Stream#$anonfun$parEvalMapActionImpl$5
[31m  RunnerCompat.scala:181                productR @ weaver.framework.RunnerCompat#$anonfun$run$12[0m


2023.03.16 00:50:39 ERROR [31m- [0mreturns a combined transaction hash for a valid transaction hex [37m[1m168ms[0m
[31m  IllegalArgumentException: Invalid point encoding 0x-35[0m

[31m  ECCurve.java:443                      org.bouncycastle.math.ec.ECCurve#decodePoint
[31m  Hex.scala:67                          org.tessellation.security.hex$Hex$Ops$newtype$#$anonfun$toPublicKeyByEC$extension$2
[31m  Hex.scala:66                          delay @ org.tessellation.security.hex$Hex$Ops$newtype$#$anonfun$toPublicKeyByEC$1
[31m  Hex.scala:65                          flatMap @ org.tessellation.security.hex$Hex$Ops$newtype$#$anonfun$toPublicKeyByEC$1
[31m  Hex.scala:60                          delay @ org.tessellation.security.hex$Hex$Ops$newtype$#toPublicKeyByEC$extension
[31m  Hex.scala:59                          flatMap @ org.tessellation.security.hex$Hex$Ops$newtype$#toPublicKeyByEC$extension
[31m  ConstructionServiceSuite.scala:355    map @ org.tessellation.rosetta.domain.construction.ConstructionServiceSuite$#$anonfun$new$62
[31m  ConstructionServiceSuite.scala:358    map @ org.tessellation.rosetta.domain.construction.ConstructionServiceSuite$#$anonfun$new$61
[31m  ConstructionServiceSuite.scala:363    map @ org.tessellation.rosetta.domain.construction.ConstructionServiceSuite$#$anonfun$new$61
[31m  ConstructionServiceSuite.scala:365    flatMap @ org.tessellation.rosetta.domain.construction.ConstructionServiceSuite$#$anonfun$new$61
[31m  RunnerCompat.scala:189                parTraverse$extension @ weaver.framework.RunnerCompat#$anonfun$run$12[0m
```

Those error messages are beyond my understanding.

cngo-github commented:
As requested, I changed the `signatures` in the schema to NEL, and I changed the signature of `combineTransation` to accept a single `RosettaSignature` instead of a NEL of `RosettaSignature`s.

cngo-github commented:
@sdavidp test tag

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +108,25 @@ object ConstructionService {
+
+    def combineTransaction(hex: Hex, signature: RosettaSignature): EitherT[F, ConstructionError, Hex] =
+      EitherT
+        .fromEither(KryoSerializer[F].deserialize[Transaction](hex.toBytes))
```
Changed.

cngo-github commented on modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:
```diff
@@ -106,5 +108,25 @@ object ConstructionService {
+            EitherT
+              .fromEither(
+                KryoSerializer[F]
+                  .serialize(Signed[Transaction](transaction, NonEmptySet.of(proof)))
```
Changed.
